### PR TITLE
frontend: only show buyable coins in moonpay dropdown

### PR DESCRIPTION
**Pull request by thisconnect**
The dropdown checks now each coin if buying is supported and
only shows those accounts.

### CODE REVIEW

File: frontends/web/src/routes/buy/info.tsx [hunk 1]
```diff
@@ -59,17 +71,31 @@ class BuyInfo extends Component<Props, State> {
...
+        Promise.all(
+            this.props.accounts.map((account) => (
+                apiGet(`account/${account.code}/exchange/moonpay/buy-supported`)
+                    .then(isSupported => (isSupported ? account : false))
+            ))
+        )
+        .then(results => results.filter(result => result))
+        // @ts-ignore
+        .then(accounts  => accounts.map(({ name, code }) => ({ text: name, value: code })))
+        .then(options => setTimeout(() => this.setState({ options }), 1000))
```
Comment (x1ddos) [inline frontends/web/src/routes/buy/info.tsx:84]: hm... why timeout?

File: frontends/web/src/routes/buy/info.tsx [hunk 2]
```diff
@@ -59,17 +71,31 @@ class BuyInfo extends Component<Props, State> {
...
 
+    private checkSupportedCoins = () => {
+        Promise.all(
+            this.props.accounts.map((account) => (
+                apiGet(`account/${account.code}/exchange/moonpay/buy-supported`)
+                    .then(isSupported => (isSupported ? account : false))
+            ))
+        )
+        .then(results => results.filter(result => result))
+        // @ts-ignore
```
Comment (x1ddos) [inline frontends/web/src/routes/buy/info.tsx:82]: what's ignored here exactly?

File: frontends/web/src/routes/buy/info.tsx [hunk 3]
```diff
@@ -170,17 +196,14 @@ class BuyInfo extends Component<Props, State> {
...
                                         },
-                                        ...accounts.map(({ code, name}) => ({
-                                            text: `${name} (${code})`,
-                                            value: code,
-                                        }))]
+                                        ...options]
                                     }
                                     onChange={e => this.setState({ selected: e.target.value})}
-                                    value={selected}
+                                    value={options.length ? selected : 'choose'}
```
Comment (x1ddos) [inline frontends/web/src/routes/buy/info.tsx:206]: if you kept it as is, `value={selected}`, would it still work?
I'm confused betwee this and the `text: t(options.length ? 'buy.info.selectPlaceholder' : 'loading')` above. seems like they're doing similar thing.

File: frontends/web/src/routes/buy/info.tsx (hunk 2 above)
Comment (thisconnect) [reply frontends/web/src/routes/buy/info.tsx:82]: So in the filter before it filters out all results that are false (the ones that returned false in buy-supported), so on the next line TypeScript still thinks there could be `false`

Similar issue "filter should narrow down output type"
https://github.com/microsoft/TypeScript/issues/20812

Couldn't manage as suggested to do "User-Defined Type Guards"
https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards

File: frontends/web/src/routes/buy/info.tsx (hunk 3 above)
Comment (thisconnect) [reply frontends/web/src/routes/buy/info.tsx:206]: `text: t(options.length ? 'buy.info.selectPlaceholder' : 'loading')` is the text of the first grayed out option, this works if there is no preselected coin.

If there is a preselected coin, i.e. 'btc', the loading message would never show as the dropdown wants to show the option with the value 'btc'. But all accounts are checked asynchronously for if they can be bought, while checking options is an empty array, and value should point to the first grayed out option with the value 'choose' .... actually if there was a coin or token called 'choose' this wouldn't work .

does that make sense?

File: frontends/web/src/routes/buy/info.tsx (hunk 2 above)
Comment (x1ddos) [reply frontends/web/src/routes/buy/info.tsx:82]: Interesting. So, something like this wouldn't work?

```ts
function isSupported(a: Account | bool): a is Account {
  return a !== false
}

.then(results => results.filter<Account>(isSupported))
```

File: frontends/web/src/routes/buy/info.tsx (hunk 3 above)
Comment (x1ddos) [reply frontends/web/src/routes/buy/info.tsx:206]: but while the accounts are being checked, there's no way to select anything regardless of whether a coin is preselected, right.

what happens if the value remains as is, `value={selected}`, and `selected` is preset to say `btc`?

File: frontends/web/src/routes/buy/info.tsx (hunk 1 above)
Comment (thisconnect) [reply frontends/web/src/routes/buy/info.tsx]: yep, there should not be a timeout, thanks for catching

File: frontends/web/src/routes/buy/info.tsx (hunk 3 above)
Comment (thisconnect) [reply frontends/web/src/routes/buy/info.tsx:206]: then it doesn't show loading, but to be honest it happens so fast, I doubt that anyone would ever see the loading text.

**Review by x1ddos — APPROVED** [review]
LGTM. Can improve later.

**Pull request merged**